python/cudf/cudf/core/frame.py

```diff
@@ -2344,83 +2349,25 @@ def replace(self, to_replace: Any, replacement: Any) -> Frame:
 
         return result
 
-    def _copy_categories(self, other, include_index=True):
-        """
-        Utility that copies category information from `other`
-        to `self`.
-        """
+    def _postprocess_columns(
+        self, other: Frame, include_index: bool = True
+    ) -> Frame:
         for name, col, other_col in zip(
             self._data.keys(), self._data.values(), other._data.values()
         ):
-            if isinstance(
-                other_col, cudf.core.column.CategoricalColumn
-            ) and not isinstance(col, cudf.core.column.CategoricalColumn):
-                self._data[name] = build_categorical_column(
-                    categories=other_col.categories,
-                    codes=as_column(col.base_data, dtype=col.dtype),
-                    mask=col.base_mask,
-                    ordered=other_col.ordered,
-                    size=col.size,
-                    offset=col.offset,
-                    null_count=col.null_count,
-                )
-        if include_index:
-            # include_index will still behave as False
-            # incase of self._index being a RangeIndex
-            if (
-                self._index is not None
-                and not isinstance(self._index, cudf.core.index.RangeIndex)
-                and isinstance(
-                    other._index,
-                    (cudf.core.index.CategoricalIndex, cudf.MultiIndex),
-                )
-            ):
-                self._index._postprocess_columns(
-                    other._index, include_index=False
-                )
-                # When other._index is a CategoricalIndex, there is
-                # possibility that corresposing self._index be GenericIndex
-                # with codes. So to update even the class signature, we
-                # have to reconstruct self._index:
-                if isinstance(
-                    other._index, cudf.core.index.CategoricalIndex
-                ) and not isinstance(
-                    self._index, cudf.core.index.CategoricalIndex
-                ):
-                    self._index = cudf.core.index.Index._from_table(
-                        self._index
-                    )
-        return self
-
-    def _copy_struct_names(self, other, include_index=True):
-        """
-        Utility that copies struct field names.
-        """
-        for name, col, other_col in zip(
-            self._data.keys(), self._data.values(), other._data.values()
-        ):
-            if isinstance(other_col, cudf.core.column.StructColumn):
-                self._data[name] = col._rename_fields(
-                    other_col.dtype.fields.keys()
-                )
+            self._data[name] = other_col._copy_type_metadata(col)
 
         if include_index and self._index is not None:
+            assert other._index is not None
```

Comment:
Should we throw with an error message here instead of assert?

Reply:
This isn't meant to ever propagate up to the user. The `assert` is used here to tell `MyPy` that `other._index` is of type `Index` (and not `Optional[Index]`):

https://mypy.readthedocs.io/en/stable/kinds_of_types.html#optional-types-and-the-none-type

Reply:
cc:  @kkraus14 for visiblity as we were just discussing this

Reply:
I think we can do better here by checking if `other` is a `DataFrameOrSeries` instead. Looking into it

Reply:
Just used a conditional instead of using an assertion here. Along the way, realized we could simplify this function further.